File: web/lib/api/endpoints/card-policy.ts
import { api } from "../client";

/** 对齐 internal/db.CardPolicy。策略跟 ICCID 走，不跟设备。 */
export interface CardPolicy {
  iccid: string;
  network_enabled: boolean;
  vowifi_enabled: boolean;
  airplane_enabled: boolean;
  ip_version: string;
  apn: string;
  /** auto = 系统推导，user = 用户显式设置 */
  source: string;
  created_at?: string;
  updated_at?: string;
}

/**
 * GET /api/cards/:iccid/policy —— 裸对象。
 * 策略不存在时后端返回 DefaultCardPolicy 而非 404，因此这里不会抛 not found。
 */
export async function getCardPolicy(iccid: string): Promise<CardPolicy> {
  return (await api.get<CardPolicy>(`/cards/${encodeURIComponent(iccid)}/policy`)).data;
}

/** PUT /api/cards/:iccid/policy —— 返回更新后的裸对象 */
export async function putCardPolicy(
  iccid: string,
  input: Partial<CardPolicy>,
): Promise<CardPolicy> {
  return (await api.put<CardPolicy>(`/cards/${encodeURIComponent(iccid)}/policy`, input)).data;
}

/** GET /api/cards/policies -> {policies} */
export async function listCardPolicies(): Promise<CardPolicy[]> {
  return (await api.get<CardPolicy[]>("/cards/policies")).data;
}

File: internal/api/card_policy.go
package api

import (
	"errors"
	"fmt"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/yuanshuai1122/vodoge/internal/db"
)

// patchCardPolicyForDevice 解析设备当前 ICCID，对 card_policies 行执行原地修改并落库。
// mutate 在 resolve 后的副本上改字段（source 会被强制为 "user"）。
// applied=false 且 err=nil 表示设备当前无 ICCID（离线/未识别），跳过落库。
func (s *Server) patchCardPolicyForDevice(deviceID string, mutate func(*db.CardPolicy)) (iccid string, applied bool, err error) {
	worker := s.pool.GetWorker(deviceID)
	if worker == nil {
		return "", false, fmt.Errorf("设备未找到")
	}
	iccid = worker.CurrentICCID()
	if iccid == "" {
		return "", false, nil
	}
	p, err := s.data().CardPolicy.Resolve(iccid)
	if err != nil {
		return iccid, false, fmt.Errorf("获取卡策略失败: %w", err)
	}
	mutate(&p)
	p.Source = "user"
	db.NormalizeCardPolicy(&p)
	if err := s.data().CardPolicy.Upsert(p); err != nil {
		return iccid, false, fmt.Errorf("保存卡策略失败: %w", err)
	}
	return iccid, true, nil
}

func (s *Server) handleGetCardPolicy(c *gin.Context) {
	iccid := c.Param("iccid")
	pol, err := s.data().CardPolicy.Get(iccid)
	if errors.Is(err, db.ErrCardPolicyNotFound) {
		// 未建档则返回默认模板（不落库，读端点保持只读语义）
		respondOK(c, db.DefaultCardPolicy(iccid))
		return
	}
	if err != nil {
		fail(c, http.StatusInternalServerError, "", err.Error())
		return
	}
	respondOK(c, pol)
}

func (s *Server) handleListCardPolicies(c *gin.Context) {
	out, err := s.data().CardPolicy.List()
	if err != nil {
		fail(c, http.StatusInternalServerError, "", err.Error())
		return
	}
	if out == nil {
		// 前端按数组消费，null 会让 .map 崩掉
		out = []db.CardPolicy{}
	}
	respondOK(c, out)
}

func (s *Server) handlePutCardPolicy(c *gin.Context) {
	iccid := c.Param("iccid")
	var req struct {
		NetworkEnabled *bool  `json:"network_enabled"`
		VoWiFiEnabled  *bool  `json:"vowifi_enabled"`
		IPVersion      string `json:"ip_version"`
		APN            string `json:"apn"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		fail(c, http.StatusBadRequest, "", err.Error())
		return
	}

	// 查出当前策略（查不到则用默认值）
	pol, err := s.data().CardPolicy.Get(iccid)
	if err != nil {
		pol = db.DefaultCardPolicy(iccid)
	}

	if req.NetworkEnabled != nil {
		pol.NetworkEnabled = *req.NetworkEnabled
	}
	if req.VoWiFiEnabled != nil {
		pol.VoWiFiEnabled = *req.VoWiFiEnabled
	}
	if req.IPVersion != "" {
		pol.IPVersion = req.IPVersion
	}
	if req.APN != "" {
		pol.APN = req.APN
	}
	pol.Source = "user"

	if err := s.data().CardPolicy.Upsert(pol); err != nil {
		fail(c, http.StatusInternalServerError, "", err.Error())
		return
	}

	respondOK(c, pol)
}
